prototypes/gclgrid/gclfmath/gclfmath.1
.TH GCLFMATH 1 "$Date: 2009/01/13 15:19:12 $"
.SH NAME
gclfmath - do simple mathematical operations to elements of GCL 3D field objects
.SH SYNOPSIS
.nf
gclfmath db gridname infield outfield [-vector -dir outdir] < commands
.fi
.SH DESCRIPTION
.LP
Several programs developed by the author utilize GCL (Geographical CurviLinear) grids
and two associated data objects called a GCLscalarfield3d and GCLvectorfield3d.  This program
can be used for fairly flexible editing of these field objects.  Applications include
building 3D velocity models, constructing an S from a P model using constant or variable
Vp/Vs ratios, zeroing components of vector fields, or applying a scale factor to fields.
The program uses the database interface to manage the different field components.  
.LP
The program is driven by edit instructions from standard input.  The format of the
lines that define the changes requested are described below.   Operations allowed are
the scalar operations of multiple, add, and assign.  The user is reminded that 
scalar division by a constant c can be done with a multiply by 1/c.  Similarly,
subtraction is addition with a negative sign.  The format allows one to reference one
grid point at time, ranges over one or more indices, or the entire grid.  
.LP
The \fIdb\fR is a database name and must be the first argument.  \fIgridname\fR 
and \fIinfile\fR are intimately linked.  The \fIgridname\fR is a key used to define
a unique grid geometry definition for the data stored in the field component tagged
with the key \fIinfield\fR.  The result of running this program will be stored in the
database with a new field tagged with the name \fIoutfield\fR.  The result by definition
will be associated with the same parent grid geometry defined by \fIgridname\fR.
.SH OPTIONS
.IP -vector
Use this flag if the field to be modified is a vector field.  The default assumes
the input is a scalar field.  
.IP -dir
Directory name into which the new field data file will be written.  
Default is editedfields.
.SH FILES
.LP
Standard input lines define the way the field is to be edited.  The 
structure of each lines is:
.nf
x1range x2range x3range vrange opcode value
.fi
where x1range, x2range, x3range, and vrange define a range of grid index values
that opcode is to be applied using the number set as value.  The range parameters 
are integer ranges in one of five possible formats:  n-m (apply to range n to m inclusive);
n- (apply to n to maximum for the associated coordinate direction; -m (apply to 0 to 
value m inclusive); n (apply only to value n); or * (apply to entire range).  
.LP
\fIExamples:\fR
.nf
0 18-29 * * = 3.5 
.fi
This would assign the value of 3.5 to x1 grid point 0, x2 ranginf from 18 to 29, and
the entire range of x3.  I the -vector flag were used all components of each vector 
would be set to 3.5.  For scalar fields the 4th column is ignored.
.nf
2- * -15 2 * 0.5
.fi
This would multiply components of the field by 0.5 that satisfy the following
conditions:  x1 from 2 to maximum, all components of x2, and x3 from 0 to 15.
If the -vector flag is set only component 2 of the 0.5 multiplier will be applied
only to component 2.  For scalar fields this is ignored and components that satisfy
the above definition will be multiplied by 0.5.  
.nf
* * * * + 1.0
.fi
Add 1.0 to every value in the field.  
.SH DIAGNOSTICS
.LP
Diagnostic messages are printed to stderr if something fails
.SH "SEE ALSO"
.nf
See documentation for the GCLgrid library found at:
http://geology.indiana.edu/pavlis/software/gclgrid/ccdoc.GCLgrid.GCLgrid.pkg.html
.fi
.SH "BUGS AND CAVEATS"
.LP
Be careful about mixing up vector and scalar fields.  Although the libarary should not
allow this there may be residual bugs that fail to trap inconsistent requests.  i.e. 
you can use a name that is associated with a vector field but if you fail to use the -vector
flag results may be silently scrambled. 
.SH AUTHOR
.nf
Gary L. Pavlis
Department of Geological Sciences
1001 East 10th Street
Indiana University
Bloomington, IN  47405
pavlis@indiana.edu
.\" $Id: gclfmath.1,v 1.1 2009/01/13 15:19:12 pavlis Exp $


pwmig/gclfmath/gclfmath.cc
#include <iostream>
#include <sstream>
#include <string>
#include "gclgrid.h"
#include "dbpp.h"
#include "seispp.h"
using namespace std;
using namespace SEISPP;

void usage()
{
	cerr << "gclfmath db gridname infield outfield [-vector -dir outdir] < commands"<<endl;
	exit(-1);
}
class RangePair
{
public:
	int low,high;
	RangePair(int l,int h){low=l;high=h;};
};
enum FDEOperator {ADD,MULTIPLY,ASSIGN};
class FieldEditData
{
public:
	int ilow,ihigh;
	int jlow,jhigh;
	int klow,khigh;
	int llow,lhigh;
	FieldEditData(BasicGCLgrid *g,char *line);
	double apply(double val);
private:
	double opvalue;
	FDEOperator opcode;
};

RangePair parse_token(string s, int length)
{
	if(s=="*")
	{
		/* For all set full range */
		return(RangePair(0,length-1));
	}
	string slow,shigh;
	RangePair result(0,0);
	int dashposition;
	dashposition=s.find_first_of("-",0);
	if(dashposition==0)
	{
		/* Land here for things like "-22" */
		result.low=0;
		shigh.assign(s,1,s.length()-1);
		result.high=atoi(shigh.c_str());
	}
	else if(dashposition<0) 
	{
		/* Land here for things with no dash */
		result.low=atoi(s.c_str());
		result.high=result.low;
	}
	else if(dashposition==(s.length()-1))
	{
		/* land here for things like "2-" */
		result.high=length-1;
		slow.assign(s,0,dashposition-1);
		result.low=atoi(s.c_str());
	}
	else
	{
		/* For things like 2-44 */
		slow.assign(s,0,dashposition-1);
		// Use length() because assign stops copy at end of string or count
		shigh.assign(s,dashposition+1,s.length()); 
		result.low=atoi(slow.c_str());
		result.high=atoi(shigh.c_str());
	}
	return(result);
}

FieldEditData::FieldEditData(BasicGCLgrid *g,char *s)
{
	string base_error("FieldEditData constructor:  ");
	GCLscalarfield3d *gsf;
	GCLvectorfield3d *gvf;
	bool vectormode(false);
	gsf=dynamic_cast<GCLscalarfield3d *>(g);
	if(gsf==NULL)
	{
		gvf=dynamic_cast<GCLvectorfield3d *>(g);
		if(gvf==NULL)
			throw SeisppError(string("FieldEditData constructor:  ")
				+ "dynamic cast failed");
		vectormode=true;
	}
	RangePair rp(0,0);
	istringstream in(s);

	string token;
	in >> token;
	int length;
	if(vectormode)
		length=gvf->n1;
	else
		length=gsf->n1;
	rp=parse_token(token,length);
	ilow=rp.low;
	ihigh=rp.high;

	in >> token;
	if(vectormode)
		length=gvf->n2;
	else
		length=gsf->n2;
	rp=parse_token(token,length);
	jlow=rp.low;
	jhigh=rp.high;
	
	in >> token;
	if(vectormode)
		length=gvf->n3;
	else
		length=gsf->n3;
	rp=parse_token(token,length);
	klow=rp.low;
	khigh=rp.high;
		
	in >> token;
	if(vectormode)
	{
		length=gvf->nv;
		rp=parse_token(token,length);
		llow=rp.low;
		lhigh=rp.high;
	}
	else
	{
		llow=0;
		lhigh=0;
	}
	string opstring;
	in >> opstring;
	if(opstring=="*")
		opcode=MULTIPLY;
	else if(opstring=="+")
		opcode=ADD;
	else if(opstring=="=")
		opcode=ASSIGN;
	else
		throw SeisppError(base_error
		 + "Illegal operation code = "
		 + opstring);
	in >> opvalue;
}

double FieldEditData::apply(double val)
{
	double result;
	switch(opcode)
	{
	case ADD:
		result=val+opvalue;
		break;
	case MULTIPLY:
		result=val*opvalue;
		break;
	case ASSIGN:
		result=val;
		break;
	default:
		cerr << "FieldEditData::apply:  illegal opcode="
			<< opcode<<endl;
		cerr << "This should not happen so program aborts"
			<<endl;
		exit(-1);
	}
	return(result);
}


/* Design:  will allow these syntax variations for commands
  1-20 example of range 1 to 20
  * use all
  55 single number 
Each line of form:
i j k l operator value
operator can be * (multiply) or + add.
*/
bool SEISPP::SEISPP_verbose(true);
int main(int argc, char **argv)
{
	int i,j,k,l;
	if(argc<5) usage();
	string dbname(argv[1]);
	string gridname(argv[2]);
	string infield(argv[3]);
	string outfield(argv[4]);
	bool vectormode(false);
	string outdir("editedfields");
	try {
		for(i=5;i<argc;++i)
		{
			string sarg(argv[i]);
			if(sarg=="-vector")
				vectormode=true;
			else if(sarg=="-dir")
			{
				++i;
				outdir=string(argv[i]);
			}
			else
				usage();
		}
		DatascopeHandle dbh(dbname,false);
		dbh.lookup("gclgdisk");
		GCLscalarfield3d *gsf;
		GCLvectorfield3d *gvf;
		int n1,n2,n3;
		gsf=NULL;
		gvf=NULL;
		if(vectormode)
		{
			gvf=new GCLvectorfield3d(dbh.db,gridname,infield);
			n1=gvf->n1;
			n2=gvf->n2;
			n3=gvf->n3;
		}
		else
		{
			gsf=new GCLscalarfield3d(dbh.db,gridname,infield);
			n1=gsf->n1;
			n2=gsf->n2;
			n3=gsf->n3;
		}
		/* Now loop over instruction lines running potentially multiple
		operations on the same data */
		FieldEditData *fed;
		char line[256];
		while(cin.getline(line,256))
		{
			if(vectormode) 
				fed=new FieldEditData(gvf,line);
			else
				fed=new FieldEditData(gsf,line);
			for(i=0;i<n1;++i)
			 for(j=0;j<n2;++j)
			  for(k=0;k<n3;++k)
			    if( (i>=fed->ilow) && (i<=fed->ihigh) 
				 && (j>=fed->jlow) && (j<=fed->jhigh)
				&& (k>=fed->klow) && (k<=fed->khigh) )
			    {
			      if(vectormode)
			      {
			        for(l=fed->llow;l<fed->lhigh;++l)
				    gvf->val[i][j][k][l]=fed->apply(gvf->val[i][j][k][l]);
			      }
			      else
			      {
			        gsf->val[i][j][k]=fed->apply(gsf->val[i][j][k]);
			      }
			    }
		}
		if(vectormode)
			gvf->dbsave(dbh.db,string(""),outdir,outfield,outfield);
		else
			gsf->dbsave(dbh.db,string(""),outdir,outfield,outfield);
			
	} catch (int ierr)
	{
		cerr << "GCLgrid error:  something threw error code = "<<ierr<<endl;
		exit(-1);
	}
}
